packages/widget-markdown/src/extension.ts

```diff
@@ -0,0 +1,120 @@
+  id: getMarkdownDocumentId(fileUri.path),
+  path: fileUri.path,
+  name: path.basename(fileUri.path),
+  content: undefined,
+})
+
+const loadMarkdownDocuments = async () => {
+  const { workspace } = vscode
+  const fileUris = await workspace.findFiles('*.md')
+  const markdownDocuments = fileUris.map(uriToMarkdownDocument)
+
+  return markdownDocuments
+}
+
+export class MarkdownExtensionManager extends ExtensionManager<State, {}> {
+  constructor (context: vscode.ExtensionContext, channel: vscode.OutputChannel) {
+    super(
+      context,
+      channel,
+      STATE_KEY,
+      {},
+      { markdownDocuments: [], selectedMarkdownContent: undefined }
+    )
+
+    // Load initial documents
+    loadMarkdownDocuments().then((markdownDocuments) => {
+      this.updateState('markdownDocuments', markdownDocuments)
+      this.broadcast({ markdownDocuments })
+      if (markdownDocuments.length > 0) {
+        this.loadMarkdownContent(markdownDocuments[0])
+      }
+    })
+
+    // keep watching for changes
+    const watcher = vscode.workspace.createFileSystemWatcher(
+      '**/*.md',
+      false,
+      true,
+      false
+    )
+    watcher.onDidCreate(this.addMarkdownDocument)
+    watcher.onDidDelete(this.removeMarkdownDocument)
+  }
+
+  loadMarkdownContent = async (doc: MarkdownDocument) => {
+    const uri = vscode.Uri.file(doc.path)
+    const selectedMarkdownContent = (
+      await vscode.workspace.fs.readFile(uri)
+    ).toString()
+    this.updateState('selectedMarkdownContent', selectedMarkdownContent)
+    this.broadcast({ selectedMarkdownContent })
+  }
+
+  private addMarkdownDocument = async (uri: vscode.Uri) => {
+    const doc = await uriToMarkdownDocument(uri)
+    const oldMarkdownDocuments = this.state.markdownDocuments
+    const updatedMarkdownDocuments = [...oldMarkdownDocuments, doc]
+    this.updateState('markdownDocuments', updatedMarkdownDocuments)
+    this.broadcast({ markdownDocuments: updatedMarkdownDocuments })
+  }
+
+  private removeMarkdownDocument = (uri: vscode.Uri) => {
+    const id = getMarkdownDocumentId(uri.path)
+    const oldMarkdownDocuments = this.state.markdownDocuments
+    const updatedMarkdownDocuments = oldMarkdownDocuments.filter(
+      (doc) => doc.id !== id
+    )
+    this.updateState('markdownDocuments', updatedMarkdownDocuments)
+    this.broadcast({ markdownDocuments: updatedMarkdownDocuments })
+  }
+}
+
+export function activate (
+  context: vscode.ExtensionContext,
+  channel: vscode.OutputChannel
+) {
+  const stateManager = new MarkdownExtensionManager(context, channel)
+  return {
+    marquee: {
+      disposable: stateManager,
+      defaultState: stateManager.state,
+      defaultConfiguration: stateManager.configuration,
+      setup: (tangle: Client<State>) => {
+        // load content when file is selected
+        tangle.whenReady().then(() => {
+          tangle.listen(
+            'markdownDocumentSelected',
+            (markdownDocumentSelectedId) => {
```

Comment:
@christian-bromann this seems to get fired every time there is any state change e.g. when [this](https://github.com/stateful/vscode-marquee/pull/85/files#diff-71345bf3090069793fdb6a0e464c65a831cb92854ddb78cc30085ebbcc1750faR85-R86) is executed.

Is there any way to listen only to `markdownDocumentSelected`?

Reply:
@MoSattler I believe this is a bug in `tangle`, you can update calls even though the update did not happen on your property

Reply:
I can work around it, I am still trying to find out the least hacky way of doing so.